scripts/generate_docs.sc

```diff
@@ -0,0 +1,147 @@
+
+val currentGitBranch = %%('git, "rev-parse", "--abbrev-ref", "HEAD").out.lines.head
+
+def listAllTagsInOrder: Vector[String] = %%('git, "tag", "-l", "--sort=version:refname").out.lines
+
+def isReleaseTag(tag: String) = tag.matches("""v[1-9]+\.\d+\.\d+""")
+
+/**
+  * get the major.minor version from a [v]major.minor.patch
+  *
+  * v1.4.11 -> 1.4
+  * 1.45.983 -> 1.45
+  *
+  * @param tag
+  * @return
+  */
+def releaseBranchVersion(tag: String) = tag.dropWhile(_.isLetter).takeWhile {
+  var reachedMinorVersion = false
+  elem => elem match {
+    case c if c.isDigit => true
+    case c if c == '.' && !reachedMinorVersion =>
+      reachedMinorVersion = true
+      true
+    case _ => false
+  }
+}
+
+def notIgnoredBranch(branchVersion: String) = !ignoredReleaseBranchesVersions.contains(branchVersion)
+
+def getTheLastTagVersion(tags: Seq[String]) = tags.last.drop(1)
+
+/**
+  * getting 3 latest release versions and corresponding latest tags
+  *
+  * example: List[(String, String)] = List(("1.3", "1.3.14"), ("1.4", "1.4.11"), ("1.5", "1.5.6"))
+  */
+val docsTargetVersions = listAllTagsInOrder
+  .filter(isReleaseTag)
+  .groupBy(releaseBranchVersion)
+  .filterKeys(notIgnoredBranch)
+  .mapValues(getTheLastTagVersion)
+  .toList.sortBy(_._1)
+  .takeRight(3)
+
+val (latestReleaseBranch, latestReleaseVersion) = docsTargetVersions.last
+
+val buildDir = makeTmpDir()
+
+val docsDir = pwd/up/"docs"
+
+val rootDocsDir = buildDir / 'docs
+
+def generateDocsForVersion(docsPath: Path, version: String, outputPath: String = "_site"): Unit = {
+  %('bundle, "exec", s"jekyll build --config _config.yml,_config.$version.yml -d $outputPath/$version/")(docsPath)
+}
+
+def branchForTag(version: String) = {
+  s"tags/v$version"
+}
+
+// step 1: copy docs/docs to the respective dirs
+
+def checkoutDocsToTempFolder() = {
+  docsTargetVersions foreach { case (releaseBranchVersion, tagVersion) =>
+    val tagName = branchForTag(tagVersion)
+    %git('checkout, s"$tagName")
+    val tagBuildDir = buildDir / releaseBranchVersion
+    mkdir! tagBuildDir
+    println(s"Copying $tagVersion docs to: $tagBuildDir")
+    cp.into(docsDir / 'docs, tagBuildDir)
+    cp.into(docsDir / '_layouts / "docs.html", tagBuildDir)
+    println(s"Docs folder for $releaseBranchVersion is copied to ${tagBuildDir / "docs"}")
+  }
+}
+
+// step 2: generate the default version (latest release tag)
+
+def generateTopLevelDocs() = {
+  %git('checkout, branchForTag(latestReleaseVersion))
+
+  println(s"Copying docs for $latestReleaseVersion into $buildDir")
+  cp.into(docsDir, buildDir)
+
+  println("Cleaning previously generated docs")
+  rm! rootDocsDir / '_site
+
+  println(s"Generating root docs for $latestReleaseVersion")
+  %("bundle", "install", "--path", s"vendor/bundle")(rootDocsDir)
+  %('bundle, "exec", s"jekyll build --config _config.yml -d _site")(rootDocsDir)
+}
+// step 3: generate docs for other versions
+def generateVersionedDocs() {
```

Comment:
Thanks for introducing the methods. Shouldn't the tmp dir be in this method somewhere?

I'm sorry. I think I get the general idea but I have a hard time following the code. So we 1. checkout each release branch and copy the docs to a temporary folder. We then generate to docs for the latest release in the main repo. 3. states that we generate the docs for the other version we've copied in step 2. However, I don't really see this happening here. Or rather I don't see how the versions are referenced. What do you think about something like this:

```scala
val versions: Seq[utils.SemVer] = Seq(...)

versions.map { version =>
   checkout(version)
   val tmpDir: Path = copyToTmpDir(...)
   (version, tmpDir)
}.map { case (version, docsPath) =>
   val docsDir: Path = generateDocs(version, docsPath)
   (version, docsDir) 
}.fold(targetDir) { (targetDir, (version, docsDir)) =>
   copy(docsDir, targetDir, version)
   targetDir
}
```

something in that manor.

Reply:
So, `buildDocs` is doing this. However, methods `checkoutDocsToTempFolder`, `generateTopLevelDocs` and `generateVersionedDocs` assume some side effects. It might be helpful to express these side effects by passing the directories as I suggested above.